src/charm.py

```diff
@@ -1,104 +1,328 @@
+        except CharmConfigInvalidError:
+            return
+        if not self._relation_created(F1_RELATION_NAME):
+            return
+        if not self._container.can_connect():
+            return
+        if not _get_pod_ip():
+            return
+        if not self._container.exists(path=BASE_CONFIG_PATH):
+            return
+        if not self._f1_requirer.f1_ip_address or not self._f1_requirer.f1_port:
+            return
+
+        if not self._k8s_privileged.is_patched(container_name=self._container_name):
+            self._k8s_privileged.patch_statefulset(container_name=self._container_name)
+        du_config = self._generate_du_config()
+        if config_update_required := self._is_du_config_up_to_date(du_config):
+            self._write_config_file(content=du_config)
+        service_restart_required = config_update_required
+        self._configure_pebble(restart=service_restart_required)
+
+        self._update_fiveg_f1_relation_data()
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Return whether a given Juju relation was created.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: Whether the relation was created.
+        """
+        return bool(self.model.relations.get(relation_name))
+
+    def _generate_du_config(self) -> str:
+        return _render_config_file(
+            gnb_name=self._gnb_name,
+            du_f1_interface_name=self._charm_config.f1_interface_name,
+            du_f1_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            du_f1_port=self._charm_config.f1_port,
+            cu_f1_ip_address=self._f1_requirer.f1_ip_address,
+            cu_f1_port=self._f1_requirer.f1_port,
+            mcc=self._charm_config.mcc,
+            mnc=self._charm_config.mnc,
+            sst=self._charm_config.sst,
+            tac=self._charm_config.tac,
+        )
+
+    def _is_du_config_up_to_date(self, content: str) -> bool:
+        """Decide whether config update is required by checking existence and config content.
+
+        Args:
+            content (str): desired config file content
+
+        Returns:
+            True if config update is required else False
+        """
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+                content=content
+        ):
+            return True
+        return False
+
+    def _config_file_is_written(self) -> bool:
+        return bool(self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"))
+
+    def _config_file_content_matches(self, content: str) -> bool:
+        if not self._container.exists(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            return False
+        existing_content = self._container.pull(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}")
+        if existing_content.read() != content:
+            return False
+        return True
```

Comment:
The method `_config_file_is_written` seems redundant, seeing that we already check if the file exist when checking the content.
```suggestion
    def _is_du_config_up_to_date(self, content: str) -> bool:
        """Decide whether config update is required by checking existence and config content.

        Args:
            content (str): desired config file content

        Returns:
            True if config update is required else False
        """
        if not self._config_file_content_matches(content=content):
            return True
        return False

    def _config_file_content_matches(self, content: str) -> bool:
        if not self._container.exists(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
            return False
        existing_content = self._container.pull(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}")
        if existing_content.read() != content:
            return False
        return True
```